editor/src/clj/editor/gui.clj

```diff
@@ -661,11 +746,190 @@
+(defmacro wrap-layout-property-edit-type
+  ([prop-sym edit-type-form]
+   {:pre [(symbol? prop-sym)]}
+   (let [edit-type-set-fn-form
+         (let [prop-kw (keyword prop-sym)
+               edit-type-set-fn-sym (symbol (str (name prop-sym) "-edit-type-set-fn"))]
+           `(fn ~edit-type-set-fn-sym [~'evaluation-context ~'node-id ~'old-value ~'new-value]
+              (layout-property-edit-type-set-impl ~'evaluation-context ~'node-id ~prop-kw ~'old-value ~'new-value nil)))]
+     `(assoc ~edit-type-form
+        :set-fn ~edit-type-set-fn-form
+        :clear-fn basic-layout-property-edit-type-clear-fn)))
+  ([prop-sym edit-type-form changes-fn-form]
+   {:pre [(symbol? prop-sym)]}
+   (let [edit-type-set-fn-form
+         (let [prop-kw (keyword prop-sym)
+               edit-type-set-fn-sym (symbol (str (name prop-sym) "-edit-type-set-fn"))]
+           `(fn ~edit-type-set-fn-sym [~'evaluation-context ~'node-id ~'old-value ~'new-value]
+              (layout-property-edit-type-set-impl ~'evaluation-context ~'node-id ~prop-kw ~'old-value ~'new-value ~changes-fn-form)))
+
+         edit-type-clear-fn-form
+         (let [edit-type-clear-fn-sym (symbol (str (name prop-sym) "-edit-type-clear-fn"))]
+           `(fn ~edit-type-clear-fn-sym [~'node-id ~'prop-kw]
+              (layout-property-edit-type-clear-impl ~'node-id ~'prop-kw ~changes-fn-form)))]
+     `(assoc ~edit-type-form
+        :set-fn ~edit-type-set-fn-form
+        :clear-fn ~edit-type-clear-fn-form))))
+
+;; SDK api
+(defmacro layout-property-edit-type
+  ([prop-sym edit-type-form]
+   `(g/constantly
+      (wrap-layout-property-edit-type ~prop-sym ~edit-type-form)))
+  ([prop-sym edit-type-form changes-fn-form]
+   `(g/constantly
+      (wrap-layout-property-edit-type ~prop-sym ~edit-type-form ~changes-fn-form))))
+
 (g/defnode GuiNode
   (inherits core/Scope)
   (inherits scene/SceneNode)
   (inherits outline/OutlineNode)
 
+  ;; Storage for layout overrides.
+  (property layout->prop->override g/Any (default {})
+            (dynamic visible (g/constantly false)))
+
+  ;; SceneNode property overrides with layout support
+  (property position types/Vec3 (default scene/default-position)
+            (dynamic edit-type (layout-property-edit-type position {:type types/Vec3}))
+            (value (layout-property-getter position))
+            (set (layout-property-setter position)))
+  (property rotation types/Vec4 (default scene/default-rotation)
+            (dynamic edit-type (layout-property-edit-type rotation properties/quat-rotation-edit-type))
+            (value (layout-property-getter rotation))
+            (set (layout-property-setter rotation)))
+  (property scale types/Vec3 (default scene/default-scale)
+            (dynamic edit-type (layout-property-edit-type scale {:type types/Vec3}
+                                 (fn [evaluation-context self prop-kw old-value new-value]
+                                   {:scale (some-> new-value scene/non-zeroify-scale)})))
```

Comment:
We sometimes need to do something special when setting values. Instead of the usual `(set (fn []))` declaration on the property, we have the ability to specify a `:changes-fn` in the `edit-type` that returns a map of `{:prop-kw prop-value}` that should be set, and the `:set-fn` in the `edit-type` will make sure it ends up in the correct place (I.e.  directly on the edited property, or inside the `:layout->prop->override` map).